### ISSUE TITLE
Conversion from "string" format not allowed on jQuery.multiDatesPicker

### ISSUE THREAD

**Issue by gotoashishgupta** [dubrox/Multiple-Dates-Picker-for-jQuery-UI#73]
This happens in dateConvert function which receives 'function' as from_format.

To resolve the issue in getDates function number case to 
<?php
                                                var o_dates = new Array(),
                                                totalDates = this.multiDatesPicker.dates[type].length;
                                                for(var i = 0; i < totalDates; i++)
                                                        o_dates[i] = dateConvert.call(
                                                                        this,
                                                                        this.multiDatesPicker.dates[type][i],
                                                                        format
                                                                );
                                                return o_dates; 
?>

**Comment by charger440sixpac** [comment]
There is a bug in dateConvert which masks this error. On line 238 -
```javascript
...
switch(from_format) {
	case 'object': break;
	case 'string': date = $.datepicker.parseDate(dateFormat, date); break;
	case 'number': date = new Date(date); break;
	default: $.error('Conversion from "'+ desired_format +'" format not allowed on jQuery.multiDatesPicker');
}
...
```
On the default line, **desired_format** should be **from_format**. Once this was corrected, the error message changes to reflect it is a _**function**_ being passed into the dateConvert function as 'date'. After looking at the function being passed, it is a method being added to the Array object in another .js file:

```javascript
...
Array.prototype.in_array = function(F) {
	var R = 0;
	if (F && this.length) {
		for (var I = 0; I<this.length; I++) {
			if (this[I]===F) {
				R = 1;
				break;
			}
		}
	}
	return R;
};
...
```
Anyone have any idea how this Array method is getting fed into the 'date' parameter for the MDP code?

**Comment by charger440sixpac** [comment]
Never mind, since arrays aren't really arrays in js, I know why. Until I can fix the Array.prototype instances in the other js file, I think I've got a working MDP after adding some code in the getDates() method to ignore functions in the multiDatesPicker.dates[type] array object.

**Comment by WolfieZero** [comment]
So I got this issue and it turns out one of the array node types is `function` so I just added bit of code by-pass that issue

    if (from_format === 'function') {
        return false;
    }

I just added that [here](https://github.com/dubrox/Multiple-Dates-Picker-for-jQuery-UI/blob/master/jquery-ui.multidatespicker.js#L208)

This was despite I wasn't passing a function. Not sure what issue this could cause but for me it works.

**Comment by yaozuwang** [comment]
@charger440sixpac so how exactly did you fix it?

**Comment by joomlaevolved** [comment]
The solution (to clear it up) is, as WolfieZero wrote to write 

```
if (from_format === 'function') {
    return false;
}
```
before **switch(from_format) {**

in **dateConvert** function.

**Comment by joomlaevolved** [comment]
furthermore I had to convert **getDates function  **

to

`getDates : function( format, type ) {
				if(!format) format = 'string';
				if(!type) type = 'picked';
				switch (format) {
					case 'object':
						return this.multiDatesPicker.dates[type];
					case 'string':
					case 'number':
						var o_dates = new Array();
						
						for(var i in this.multiDatesPicker.dates[type]){
							
							var dateConverted = dateConvert.call(
									this, 
									this.multiDatesPicker.dates[type][i], 
									format
								)
							if(dateConverted)	
							o_dates.push(
								dateConverted
							);
							}
							alert(o_dates.length);
								alert(JSON.stringify(o_dates));
						return o_dates;
					
					default: $.error('Format "'+format+'" not supported!');
				}
			}`

**Comment by charger440sixpac** [comment]
My modified getDates() method -

```
getDates : function( format, type ) {
	if (!format) format = 'string';
	if (!type) type = 'picked';
	switch (format) {
		case 'object':
			return this.multiDatesPicker.dates[type];

		case 'string':
		case 'number':
			var o_dates = new Array();
			if (this.multiDatesPicker.dates[type].length) {		// ADD
				for(var i in this.multiDatesPicker.dates[type]) {
					if (typeof this.multiDatesPicker.dates[type][i] != 'function') {	// ADD
						o_dates.push(
							dateConvert.call(
								this, 
								this.multiDatesPicker.dates[type][i], 
								format
							)
						);
					}	// ADD
				}
			}	// ADD
			return o_dates;

		default: $.error('Format "'+format+'" not supported!');
	}
},
```
I only corrected the variable name error in dateConverted as mentioned previous; the trapping of functions is done here before dateConverted is called.

**Comment by kefahB** [comment]
Hello everyone,
After long research, I discovered that the problem is not the viennt multiDatesPicker, we must look if a portion of the js code before multiDatePicker, I had a Array.prototype just before what made the bug then do the testes and comment out the js code that runs before I multiDatesPicker ...

I m sorry is google translate !

Bonjour à tous,
Après long recherches, j'ai découvert que le problème ne viennt pas du multiDatesPicker, il faut regarder s'il y a une portion du code js avant le multiDatePicker, moi j'avais un Array.prototype juste avant ce qui a fait le bug, alors faites des testes et mettez en commentaires les codes js qui s’exécute avant je multiDatesPicker...

Bien à vous

**Comment by danielowentorrey** [comment]
This error is fixed in version 1.6.6. I suggest that this issue be closed.
